exchanges/gateio/gateio_websocket.go

```diff
@@ -38,697 +67,904 @@ func (g *Gateio) WsConnect() error {
+			crossMarginBalanceChannel,
+			marginBalancesChannel,
+			spotBalancesChannel}...)
 	}
-	for i := range channels {
-		for j := range enabledCurrencies {
+	var subscriptions []stream.ChannelSubscription
+	var pairs []currency.Pair
+	var err error
+	for i := range channelsToSubscribe {
+		switch channelsToSubscribe[i] {
+		case marginBalancesChannel:
+			pairs, err = g.GetEnabledPairs(asset.Margin)
+		case crossMarginBalanceChannel:
+			pairs, err = g.GetEnabledPairs(asset.CrossMargin)
+		default:
+			pairs, err = g.GetEnabledPairs(asset.Spot)
+		}
+		if err != nil {
+			return nil, err
+		}
+		for j := range pairs {
 			params := make(map[string]interface{})
-			if strings.EqualFold(channels[i], "depth.subscribe") {
-				params["limit"] = 30
-				params["interval"] = "0.1"
-			} else if strings.EqualFold(channels[i], "kline.subscribe") {
-				params["interval"] = 1800
+			switch channelsToSubscribe[i] {
+			case spotOrderbookChannel:
+				params["level"] = 100
+				params["interval"] = kline.HundredMilliseconds
+			case spotCandlesticksChannel:
+				params["interval"] = kline.FiveMin
+			case spotOrderbookUpdateChannel:
+				params["interval"] = kline.ThousandMilliseconds
 			}
-
-			fpair, err := g.FormatExchangeCurrency(enabledCurrencies[j],
-				asset.Spot)
+			fpair, err := g.FormatExchangeCurrency(pairs[j], asset.Spot)
 			if err != nil {
 				return nil, err
 			}
 
 			subscriptions = append(subscriptions, stream.ChannelSubscription{
-				Channel:  channels[i],
+				Channel:  channelsToSubscribe[i],
 				Currency: fpair.Upper(),
 				Params:   params,
-				Asset:    asset.Spot,
 			})
 		}
 	}
 	return subscriptions, nil
 }
 
-// Subscribe sends a websocket message to receive data from the channel
-func (g *Gateio) Subscribe(channelsToSubscribe []stream.ChannelSubscription) error {
-	payloads, err := g.generatePayload(channelsToSubscribe)
+// handleSubscription sends a websocket message to receive data from the channel
+func (g *Gateio) handleSubscription(event string, channelsToSubscribe []stream.ChannelSubscription) error {
+	payloads, err := g.generatePayload(event, channelsToSubscribe)
 	if err != nil {
 		return err
 	}
-
 	var errs common.Errors
 	for k := range payloads {
-		resp, err := g.Websocket.Conn.SendMessageReturnResponse(payloads[k].ID, payloads[k])
+		err = g.Websocket.Conn.SendJSONMessage(payloads[k])
 		if err != nil {
 			errs = append(errs, err)
 			continue
 		}
-		var response WebsocketAuthenticationResponse
-		err = json.Unmarshal(resp, &response)
-		if err != nil {
-			errs = append(errs, err)
-			continue
-		}
-		if response.Result.Status != "success" {
-			errs = append(errs, fmt.Errorf("%v could not subscribe to %v",
-				g.Name,
-				payloads[k].Method))
-			continue
+		channel := make(chan *WsEventResponse)
+		g.WsChannelsMultiplexer.Register <- &wsChanReg{
+			ID:   strconv.FormatInt(payloads[k].ID, 10),
+			Chan: channel,
+		}
+		ticker := time.NewTicker(time.Second * 3)
```

Comment:
```suggestion
		t := time.NewTicker(time.Second * 3)
```
As `ticker` conflicts with package name